File: 3362-zero-array-transformation-iii/3362-zero-array-transformation-iii.cpp
class Solution {
public:
    int maxRemoval(vector<int>& nums, vector<vector<int>>& queries) {
        int m = queries.size(), n = nums.size();
        sort(queries.begin(), queries.end());
        priority_queue<int> heap;
        vector<int> deltaArray(n + 1);
        int j = 0, operations = 0;

        for (int i = 0; i < n; ++i){
            operations += deltaArray[i];

            while (j < m && queries[j][0] == i){
                heap.push(queries[j][1]);
                ++j;
            }
            
            while (operations < nums[i] && !heap.empty() && heap.top() >= i){
                ++operations;
                --deltaArray[heap.top() + 1];
                heap.pop();
            }

            if (operations < nums[i]) return -1;
        }

        return heap.size();
    }
};

File: 3362-zero-array-transformation-iii/3362-zero-array-transformation-iii.py
class Solution:
    def maxRemoval(self, nums: List[int], queries: List[List[int]]) -> int:
        queries.sort()
        heap, m, n = [], len(queries), len(nums)
        delta_arr = [0] * (n + 1)
        ops, j = 0, 0

        for i, num in enumerate(nums):
            ops += delta_arr[i]

            while j < m and queries[j][0] == i:
                heappush(heap, -queries[j][1])
                j += 1
            
            while ops < num and heap and -heap[0] >= i:
                ops += 1
                delta_arr[-heappop(heap) + 1] -= 1

            if ops < num:
                return -1

        return len(heap)
